src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java

```diff
@@ -31,17 +32,18 @@
  * @author Kohsuke Kawaguchi
  */
 public class SubmoduleOption extends GitSCMExtension {
-    /**
-     * Use --recursive flag on submodule commands - requires git>=1.6.5
-     * Use --remote flag on submodule update command - requires git>=1.8.2
-     * Use --reference flag on submodule update command - requires git>=1.6.4
-     */
     private boolean disableSubmodules;
+    /** Use --recursive flag on submodule commands - requires git>=1.6.5 */
     private boolean recursiveSubmodules;
+    /** Use --remote flag on submodule update command - requires git>=1.8.2 */
     private boolean trackingSubmodules;
+    /** Use --reference flag on submodule update command - requires git>=1.6.4 */
     private String reference;
     private boolean parentCredentials;
     private Integer timeout;
+    /** Use --depth flag on submodule update command - requires git>=1.8.4 */
+    private boolean shallow;
+    private int depth = 1;
```

Comment:
I think that the default value for `depth` should be 0 to perform a full depth fetch as was done before.  We don't want to change the way submodules are cloned by default.

Reply:
The default value of 1 in `SubmoduleOption` originates from the one ([here](https://github.com/jenkinsci/git-plugin/blob/git-4.0.0-beta3/src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java#L29)) in `CloneOption`.
The only difference is that in `CloneOption` there is an extra check ([here](https://github.com/jenkinsci/git-plugin/blob/git-4.0.0-beta3/src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java#L117)) that applies the `depth` only if `shallow` is `true`. This is not really required as exactly this check is also performed in the git-client-plugin in `FetchCommand`, `CloneCommand` and `SubmoduleUpdateCommand`. It is even stated in their Javadocs that `depth` has no effect if `shallow` is `false`.

I would therefore like to leave it as is.

Reply:
The default value of 1 in `SubmoduleOption` originates from the one ([here](https://github.com/jenkinsci/git-plugin/blob/git-4.0.0-beta3/src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java#L29)) in `CloneOption`.
The only difference is that in `CloneOption` there is an extra check ([here](https://github.com/jenkinsci/git-plugin/blob/git-4.0.0-beta3/src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java#L117)) that applies the `depth` only if `shallow` is `true`. This is not really required as exactly this check is also performed in the git-client-plugin in `FetchCommand` ([here](https://github.com/jenkinsci/git-client-plugin/blob/git-client-3.0.0-beta5/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java#L399)), `CloneCommand` ([here](https://github.com/jenkinsci/git-client-plugin/blob/git-client-3.0.0-beta5/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java#L627-L632) - uses fetch) and `SubmoduleUpdateCommand` ([here](https://github.com/jenkinsci/git-client-plugin/blob/git-client-3.0.0-beta5/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java#L1155)). It is even stated in their Javadocs ([here](https://github.com/jenkinsci/git-client-plugin/blob/git-client-3.0.0-beta5/src/main/java/org/jenkinsci/plugins/gitclient/FetchCommand.java#L62), [here](https://github.com/jenkinsci/git-client-plugin/blob/git-client-3.0.0-beta5/src/main/java/org/jenkinsci/plugins/gitclient/CloneCommand.java#L112), [here](https://github.com/jenkinsci/git-client-plugin/blob/git-client-3.0.0-beta5/src/main/java/org/jenkinsci/plugins/gitclient/SubmoduleUpdateCommand.java#L73)) that `depth` has no effect if `shallow` is `false`.

I would therefore like to leave it as is.

Reply:
Thanks for the clarification.  Agreed.

Reply:
I just opened pull request #614 which is based on this one for further improvements.

Reply:
I performed some more testing.

Result: I will already add the same `depth` check in `SubmoduleOption` as already exists in `CloneOption`.
```
if (shallow && depth > 1) {
    cmd.depth(depth);
}
```
Otherwise if the `depth` field is left empty in the UI, its value gets 0 and updating submodules with `--depth=0` does not perform anything.

Unfortunately no error is reported currently. This is due to pull request jenkinsci/git-client-plugin#348.

Reply:
Should we revert #348 or do we want to give some time in hopes the submitter will make the necessary corrections?

Reply:
It should really be fixed before any release.

I would wait for @nre-ableton to provide a fix. I already provided him some hints.

@MarkEWaite please update the pull request reference to *git-client-plugin*. Currently it references a pull request in *git-plugin* which is confusing.